File: leet-code/169. Majority Element/MajorityElement169.java
package com.example;

import java.util.HashMap;
import java.util.Map;

public class MajorityElement169 {

    public int majorityElement(int[] nums) {
        Map<Integer, Integer> cnt = new HashMap<>();
        for (int num : nums) cnt.put(num, cnt.getOrDefault(num, 0) + 1);
        int majority = 0;
        for (int num : nums) {
            if (cnt.get(num) >= (nums.length + 1) / 2) {
                majority = num;
                break;
            }
        }
        return majority;
    }
}

File: leet-code/169. Majority Element/MajorityElement169.kt
package com.example

class MajorityElement169 {
    fun majorityElement(nums: IntArray): Int {
        return findMajority(nums, 0, nums.size - 1)
    }

    private fun findMajority(nums: IntArray, left: Int, right: Int): Int {
        if (left >= right) {
            return nums[left]
        }

        val mid = left + (right - left) / 2
        val m1 = findMajority(nums, left, mid)
        val m2 = findMajority(nums, mid + 1, right)
        var countM1 = 0
        for (i in left..right) {
            if (nums[i] == m1) {
                countM1++
            }
        }
        return if (countM1 > (right - left + 1) / 2) {
            m1
        } else {
            m2
        }
    }
}
